src/vm/costs/mod.rs

```diff
@@ -194,6 +260,235 @@ pub enum CostErrors {
     CostContractLoadFailure,
 }
 
+fn load_state_summary(clarity_db: &mut ClarityDatabase) -> Result<CostStateSummary> {
+    let last_processed_at = match clarity_db.get_value(
+        "vm-costs::last-processed-at-height",
+        &TypeSignature::UIntType,
+    ) {
+        Some(v) => u32::try_from(v.expect_u128()).expect("Block height overflowed u32"),
+        None => return Ok(CostStateSummary::empty()),
+    };
+
+    let metadata_result = clarity_db
+        .fetch_metadata_manual::<String>(
+            last_processed_at,
+            &STACKS_BOOT_COST_VOTE_CONTRACT,
+            "state_summary",
+        )
+        .map_err(|e| CostErrors::CostComputationFailed(e.to_string()))?;
+    let serialized: SerializedCostStateSummary = match metadata_result {
+        Some(serialized) => serde_json::from_str(&serialized).unwrap(),
+        None => return Ok(CostStateSummary::empty()),
+    };
+    Ok(CostStateSummary::from(serialized))
+}
+
+fn store_state_summary(
+    clarity_db: &mut ClarityDatabase,
+    to_store: &CostStateSummary,
+) -> Result<()> {
+    let block_height = clarity_db.get_current_block_height();
+
+    clarity_db.put(
+        "vm-costs::last-processed-at-height",
+        &Value::UInt(block_height as u128),
+    );
+    let serialized_summary =
```

Comment:
Pretty sure this is the case, but just confirming -- metadata stored for a contract is _not_ consensus-critical, right?  It's just a (block-hash, key, value) triple in the side-store, with no direct link to the MARF, right?  Otherwise we'll want to avoid using JSON here.

Reply:
Yep, metadata isn't consensus critical.